designs/disruption-controls.md

```diff
@@ -0,0 +1,118 @@
+# Disruption Controls - (in v1beta1+)
+
+## Motivation
+Users have been asking for more control around the speed of Consolidation and the ability to restrict disruptions to their nodes. The demand for this is reflected by the +1s in [#1738](https://github.com/aws/karpenter/issues/1738). This is increasingly important as clusters can run at a highly varied scales, where some users have reported scale-down behaving too slow at higher scales, or too quick at smaller scales. Additionally, users that want to control when nodes can be disrupted must add PDBs to block node disruptions, add `do-not-evict` annotations on pods for sensitive workloads, or add `do-not-consolidate` annotations on nodes that shouldn't be consolidated. Natively integrating a better way to ensure nodes aren't disrupted will ease the burden that users have to take on with disruption.
+
+These controls will include a new `Disruption` block in the `v1beta1` `NodePool` that will contain all scale-down behavioral fields. Initially, this will include `Budgets`, which define a (1) parallelism of how many nodes can be deprovisioned at a time and a (2) cron schedule that determines when the parallelism applies, and a `ConsolidateAfter` field that will allow users to affect the speed at which Karpenter will scale down underutilized nodes. Future scale-down behavioral fields should be colocated with these fields.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    budgets:
+    # On Weekdays during business hours, don't do any deprovisioning.
+    - crontab: "0 9 * * mon-fri"
```

Comment:
Would drift be controlled by this disruption budget?

We've found another pain point for management is that we have some provisioners that scale up/down regularly, so we'd like to try and let them drain out naturally (in this case, they have expensive to rebuild caches that we'd like to not force a rebuild of).

At the same time, we're ok with forcing drifting if we can control when it happens.

Reply:
Yep! I'm working on pulling out some of the synchronous waits we do in code to enable Karpenter to drift/expire/consolidate multiple nodes at a time. https://github.com/aws/karpenter-core/issues/255 is where I'm tracking this. 

The maxUnavailable will restrict all disruption mechanisms and include any other terminating nodes that may have been done manually in calculations.

Reply:
I guess we were more interested in the ability to control when drift happens, either through this budget mechanism and/or by Provisioner, rather than speeding up the process.

Reply:
Drift will happen any time you change the resolved resources in the cloud provider or change the CRDs. Is Karpenter drifting too often for you? If you set the Budget Crontab to restrict disruptions at times where you don't want drift to happen, you can still set a disruption budget for any other time where the previous crontab isn't active.

Reply:
Unfortunately, we have quite a few workload types that are singletons or cannot handle any disruption. Or have expensive caches that we'd like to not disrupt during the day. Being able to use the budget crontab would thus be a nice operational improvement. 

Generally, we roll changes to the AMI or Provider and just wanted more control over how it gets rolled out.

Reply:
We'd discussed scoping specific budgets for specific types of actions. For example, you could constrain drift on weekends to 0, but allow consolidation of 1 at a time.

Reply:
At least for the discussions we've had, just being able to fit drift into a schedule was fine. Our current issue was just that there was no control for drift.

Reply:
And to be clear, you'd only want to constrain drift, not other kinds of deprovisioning at that time.

Reply:
I think we'd like the ability, Drift is just top of mind as currently there is no way to constrain it beyond global disable/enable where as a lot of the other options you can control it through things like multiple provisioners.